### DISCUSSION TITLE
Fan toggle (M8 and M9) doesn't work

### DISCUSSION THREAD

**Discussion by Momboz52** [arkypita/LaserGRBL#2054] [General]
This function exists in the last version of the LaserGRBL Customs Buttons (CustomButtons.gz) from the web site and the toggle fan doesn't react.
Any advice ?

**Comment by arkypita** [comment]
This feature should be implemented by the controller, lasergrbl simply send the standard command that is intended to turn on off the coolant but it is responsibility of the controller to implement it.

If it is not implemented by the controller, there is no way for lasergrbl to control it.

**Comment by Momboz52** [comment]
Many thanks for your feedback.
The case is indeed a controller matter. However, the lack of these M8 and M9 implementations is done on purpose according to Atomstack, the engraver manufacturer. The reason for that is that in the Atomstack laser systems the fan must be running as soon as the device is powered on for cooling purposes. Then I can't understand the implementation of this custom button (toggle fan) in case of use of LaserGRBL for driving laser machines. This could even cause severe damages to the engraver in case the fan is turned off.

**Comment by StuartB4** [comment]
It's not for the laser fan. It can be used to toggle a seperate fan for exhaust or extra cooling, or something like air assist or even lighting around the inside of a case, if your laser is encased. 
When you enable M8 it just sends 5 volts to what ever is connected, to switch it on. 
M9 sends 0 volts to switch it off. 

The 5 volts can be used to switch a transistor for low power equipment or a relay to allow you to run higher powered stuff. 
It's used in CNC machines for switching the liquid coolant on and off. But as you don't use liquid coolant on a laser engraver you can use it for something else.

I did a quick video with a multimeter connected to check the voltage.
As soon as the engraving starts and M8 is read from the gcode the signal from the coolant pin on my controller board sends out 5 volts.
At the end it reads M9 and switches the voltage off. 

https://youtu.be/KIuOpZEc-7o

**Comment by Momboz52** [comment]
Many thanks for the clarification and for the annexed video.
I did misunderstand things because of the naming “Toggle Fan ON/OFF”. I would have named it “Additional output” or something similar.
By the way, I can't see any additional output on my Atomstack P7 M40 that I could control with this G-Code messages M8 and M9 and these are not implemented anyway by the manufacturer.
Once again many thanks.

**Reply by StuartB4** [reply]
You're welcome. 
A lot of the newer off the shelf engravers now don't have the pins or output sockets for other things. Mine is a home made machine using an Arduino UNO as the main controller, that has the control code (GRBL) programmed in to it, and that connects to a seperate CNC shield with all the output pins already there, them pins are connected through the Arduino when they are plugged together. It also has the stepper motor drivers on it which are replaceable if one goes faulty.
